Chapters/Common/StringAlgorithms.cs

```diff
@@ -0,0 +1,315 @@
+        }
+        
+        //4
+        public void RemoveSpacesBeforePunctuation()
+        {
+            for (int i = 1; i < Input.Length; i++)
+                if (Input[i] == ' ' && char.IsPunctuation(Input[i - 1]))
+                    Input.Remove(i, 1);
+        }
+        
+        //5
+        public void InsertNumberAfterSpaces(int num)
+        {
+            for (int i = 0; i < Input.Length - 1; i++)
+                if (Input[i] == ' ')
+                    Input.Insert(i + 1, num.ToString());
+        }
+        
+        //6
+        public void InsertNumberBeforeExclamation(int num)
+        {
+            for (int i = 1; i < Input.Length; i++)
+                if (Input[i] == '!')
+                    Input.Insert(i, num.ToString());
+        }
+
+        //7
+        public int SumAllNumbers()
+        {
+            int step = 1;
+            int sum = 0;
+            int start = 0;
+            bool isFirst = false;
+
+            for (int i = 0; i < Input.Length; i+=step)
+            {
+
+                if (char.IsDigit(Input[i]) && !isFirst)
+                {
+                    isFirst = true;
+                    start = i;
+                }
+                else
+                {
+                    isFirst = false;
+                    step = i - start;
+                    
+                    byte[] num = new byte[step];
+                    for (int j = start; j < i; j++)
+                        num[i - j] = Convert.ToByte(Input[j]);
+                    for (int j = num.Length - 1; j > 0; j--)
+                        sum += num[j] * (int)Math.Pow(10, num.Length - 1 - j);
+                }
+                
+            }
+            return sum;
+        }
+
+        //8
+        public void RemoveNumbers()
+        {
+            for (int i = 0; i < Input.Length; i++)
+                if (char.IsDigit(Input[i]))
+                    Input.Remove(i, 1); 
+        }
+
+        //9
+        public int SumAllSimpleNumbers()
+        {
+            int step = 1;
+            int sum = 0;
+            int start = 0;
+            int number = 0;
+            bool isFirst = false;
+
+            for (int i = 0; i < Input.Length; i += step)
+            {
+
+                if (char.IsDigit(Input[i]) && !isFirst)
+                {
+                    isFirst = true;
+                    start = i;
+                }
+                else
+                {
+                    isFirst = false;
+                    step = i - start;
+
+                    byte[] num = new byte[step];
+                    for (int j = start; j < i; j++)
+                        num[i - j] = Convert.ToByte(Input[j]);
+                    for (int j = num.Length - 1; j > 0; j--)
+                        number += num[j] * (int)Math.Pow(10, num.Length - 1 - j);
+                    if (BinaryFile.IsSimple(number))
+                        sum += number;
+                }
+            }
+            return sum;
```

Comment:
I've seen some similar code, try to make some universal method for these purposes.